pkg/action/executor/machine_scalers.go

```diff
@@ -0,0 +1,370 @@
+	}
+	return nil
+}
+
+// checkMachineSuccess checks whether machine has been created successfully.
+func (controller *machineDeploymentController) checkMachineSuccess(args ...interface{}) (bool, error) {
+	machineName := args[0].(string)
+	machine, err := controller.request.client.machine.Get(machineName, metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+	if machine.ObjectMeta.CreationTimestamp.String() != "" && machine.Status.ErrorMessage == nil {
+		return true, nil
+	}
+	return false, nil
+}
+
+// isNodeReady checks whether the node is ready.
+func (controller *machineDeploymentController) isMachineReady(args ...interface{}) (bool, error) {
+	machineName := args[0].(string)
+	machine, err := controller.request.client.machine.Get(machineName, metav1.GetOptions{})
+	if err != nil {
+		return false, err
+	}
+	if machine.Status.ErrorMessage != nil {
+		return true, nil
+	}
+	return true, nil
+}
+
+// waitForNodeProvisioning waits for the new node to be provisioned with timeout.
+func (controller *machineDeploymentController) waitForNodeProvisioning(newMachine *clusterv1.Machine) error {
+	descr := fmt.Sprintf("machine %s Node creation status is final", newMachine.Name)
+	err := controller.waitForState(descr, controller.checkMachineSuccess, newMachine.Name)
+	if err != nil {
+		return err
+	}
+	machine, err := controller.request.client.machine.Get(newMachine.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	if machine.Status.ErrorMessage != nil {
+		err = fmt.Errorf("machine %s failed to create new Node: %v: %s",
+			newMachine.Name, newMachine.Status.ErrorReason, *newMachine.Status.ErrorMessage)
+		return err
+	}
+	newNName := newMachine.ObjectMeta.Name
+	// wait for new Node to be in Ready state
+	descr = fmt.Sprintf("machine %s is Ready", newNName)
+	return controller.waitForState(descr, controller.isMachineReady, newNName)
+}
+
+// isNodeDeletedOrNotReady checks whether the node is deleted or not ready.
+func (controller *machineDeploymentController) isNodeDeletedOrNotReady(args ...interface{}) (bool, error) {
+	machineName := args[0].(string)
+	machine, err := controller.request.client.machine.Get(machineName, metav1.GetOptions{})
+	if err != nil {
+		return true, nil
+	}
+	if machine.ObjectMeta.DeletionTimestamp != nil {
+		return true, nil
+	}
+	return false, nil
+}
+
+// waitForNodeDeprovisioning waits for the new node to be de-provisioned with timeout.
+func (controller *machineDeploymentController) waitForNodeDeprovisioning(machine *clusterv1.Machine) error {
+	deletedNName := machine.Name
+	descr := fmt.Sprintf("node %s deleted or exited Ready state", deletedNName)
+	return controller.waitForState(descr, controller.isNodeDeletedOrNotReady, deletedNName)
+}
+
+// waitForState Is the function that allows to wait for a specific state, or until it times out.
+func (controller *machineDeploymentController) waitForState(stateDesc string, f stateCheck, args ...interface{}) error {
```

Comment:
Is it possible to reuse the existing `RetrySimple` or `RetryDuring` function defined in `go_util.go` file?

Reply:
The semantics are subtly different. Here I'm waiting for the state to happen, rather than retry several times. I could shoehorn it in, but the result will likely be more confusing, sorry.